synapse/logging/context.py

```diff
@@ -434,12 +436,18 @@ def stop(self, rusage: "Optional[resource._RUsage]") -> None:
             self._resource_usage.ru_utime += utime_delta
             self._resource_usage.ru_stime += stime_delta
 
-            # if we have a parent, pass our CPU usage stats on
+            # if we have a parent, pass the change in our usage stats on.
             if self.parent_context:
-                self.parent_context._resource_usage += self._resource_usage
-
-                # reset them in case we get entered again
-                self._resource_usage.reset()
+                if self._prev_resource_usage:
+                    delta_resource_usage = (
+                        self._resource_usage - self._prev_resource_usage
+                    )
+                else:
+                    delta_resource_usage = self._resource_usage
+                self.parent_context._resource_usage += delta_resource_usage
+
+                # Store the current usage for future deltas.
+                self._prev_resource_usage = self._resource_usage.copy()
```

Comment:
this feels somewhat overcomplex. It's worth remembering this is one of the hottest hot paths in synapse, so I'd rather avoid having to maintain two copies of the resource_usage etc.

why don't we:

```
self._record_cputime(utime_delta, stime_delta)

def _record_cputime(self, utime_delta, stime_delta):
    self._resource_usage.ru_utime += utime_delta
    self._resource_usage.ru_stime += stime_delta
    if self.parent_context:
        self.parent_context._record_cputime(utime_delta, stime_delta)
```

and update `add_database_transaction` to recurse up the parent list too?

Reply:
Ah, that sounds like a better solution -- I wasn't seeing the recursive solution for the transactions. 👍